Kwenta/kwenta
# EIP 1559 support for transactions - mainnet

## Description
In order to support eip 1559 transactions on mainnet, the gas configuration should take in an object that includes the following:

```
{
    maxPriorityFeePerGas,
    maxFeePerGas
}
```

on all other networks (ie optimism) the gas object should take the following configuration:

```
{
    gasPrice
}
```

by adding this functionality to the `useGas` hook, the transactions that get queued in the app should be able to support both 1559 and non 1559 transactions. Keep in mind that users can set their own `customGasPrice` in the app.

## Related issue
https://github.com/Kwenta/kwenta/issues/361

## Motivation and Context

refactoring gas logic into a `useGas` hook and setting up kwenta transactions to support EIP-1559.

I have created an object that can be used for gas configurations called `gasConfig`. you can see this being swapped in the transactions across these Kwenta components:

- [ManageShort](https://github.com/Kwenta/kwenta/pull/377/files#diff-eb24848c2ea707e624eaa2197451034fd73182e74674eef499cf6dabcf60f09bR369)
- [useExchange#1](https://github.com/Kwenta/kwenta/pull/377/files#diff-775118eccd3373aaed2a2ec3d37976b257a4fea20f41dc649d70b962b253674dR674)
- [useExchange#2](https://github.com/Kwenta/kwenta/pull/377/files#diff-775118eccd3373aaed2a2ec3d37976b257a4fea20f41dc649d70b962b253674dR723)
- [ManageShortAction](https://github.com/Kwenta/kwenta/pull/377/files#diff-eb24848c2ea707e624eaa2197451034fd73182e74674eef499cf6dabcf60f09bR369)

the main changes for this pr are in the `useGas` hook [here](https://github.com/Kwenta/kwenta/pull/377/files#diff-9b34fa921e4ee318e5ee18c3c3da462935b3f3a00188ce46bc41c66f5585d60aR31-R61). I refactored all the consistent gas related items into the following object for useGas:

```
const useGas = () => {
  ...
  return {
    gasPrice,
    gasPriceWei,
    getGasLimitEstimate,
    maxFeePerGas,
    maxPriorityFeePerGas,
    gasPrices,
    gasSpeed,
    setGasSpeed,
    isCustomGasPrice,
    customGasPrice,
    setCustomGasPrice,
    gasConfig,
  }
}
```

## How Has This Been Tested?

** I repeated the same process described below on `dev` and this branch.

- on mainnet:
1) visit the exchange
2) select two currencies to exchange and input a value
3) select `Submit Order`
4) inspected the transaction to make sure a relatively appropriate `Max Base Fee` and `Max Priority Fee` are set

## Screenshots (if appropriate):

### BEFORE changes on `dev` branch(mainnet):

![Screen Shot 2022-03-02 at 8 38 43 AM](https://user-images.githubusercontent.com/5998100/156395039-72e8311b-8cb1-40ae-ba66-b262eced24c0.png)
![Screen Shot 2022-03-02 at 8 38 59 AM](https://user-images.githubusercontent.com/5998100/156395046-11bd1751-5a00-4b55-a424-19af1a1b2e52.png)

### AFTER changes on this branch(mainnet):

![Screen Shot 2022-03-02 at 8 40 47 AM](https://user-images.githubusercontent.com/5998100/156395444-4fdfe187-f8fb-4908-bdd9-8b9ea7d17b10.png)
![Screen Shot 2022-03-02 at 8 40 57 AM](https://user-images.githubusercontent.com/5998100/156395450-babc9a2e-c9d7-43f0-a2ac-9dc9f50b412c.png)

### BEFORE changes on `dev` branch(optimism):

![Screen Shot 2022-03-02 at 8 37 07 AM](https://user-images.githubusercontent.com/5998100/156395678-2368d7e5-10ae-4ad0-8c32-0049b248b45f.png)

### AFTER changes on this branch(optimism):
![Screen Shot 2022-03-02 at 8 43 43 AM](https://user-images.githubusercontent.com/5998100/156395975-f7766938-caee-4032-924e-97efb3057024.png)
![Screen Shot 2022-03-02 at 8 43 51 AM](https://user-images.githubusercontent.com/5998100/156395980-c3a576af-d70d-4bd6-b278-037cb54bc5c9.png)
